src/AAL/AAL_knapsack.cpp

```diff
@@ -0,0 +1,78 @@
+#include "AAL_knapsack.h"
+
+#include <algorithm>
+#include <vector>
+
+class Subset {
+private:
+  float p_size;
+  float p_value;
+  std::vector<unsigned int> p_set;
+
+public:
+  Subset() {
+    p_size  = 0;
+    p_value = 0;
+    p_set.resize(0);
+  }
+  void addItem(const float size, const float value, const unsigned int item) {
+    p_size += size;
+    p_value += value;
+    p_set.push_back(item);
+  }
+
+  float size() const {
+    return this->p_size;
+  }
+
+  float value() const {
+    return this->p_value;
+  }
+
+  unsigned int numElements() const {
+    return p_set.size();
+  }
+
+  unsigned int element(const unsigned int i) const {
+    return p_set[i];
+  }
+};
+
+AAL_knapsackResult AAL_knapsack_exact(AAL_knapsackInstance instance) {
+  Subset emptyset = Subset();
+  std::vector<Subset> subsets{emptyset};
+  for (unsigned int i = 0; i < instance.numberOfItems; ++i) {
+    const unsigned int subsetsSize = subsets.size();
+    for (unsigned int j = 0; j < subsetsSize; ++j) {
+      // if there is enough space left add the item
+      if (subsets[j].size() + instance.itemSizes[i] <= instance.capacity) {
+        Subset newSubset = subsets[j];
+        newSubset.addItem(instance.itemSizes[i], instance.itemValues[i], i);
+        subsets.push_back(newSubset);
+      }
+      else {
+        break;
+      }
+    }
+    std::sort(subsets.begin(), subsets.end(), [](Subset a, Subset b) { return a.size() < b.size(); });
```

Comment:
I think here is still potential for an improvement of performance by sorting just pointers to the subsets.

Reply:
Yes, since you need the size aswell, I guess tuples of pointer + size are pretty good. We don't want to lookup some distant memory everytime just to get the size.

Reply:
I'm not yet sure about what you mean by +size. Do you suggest to have an extra array for the sizes?

Reply:
Yes, an array where each entry contains a pointer and a size where we sort by size and the pointer tells us what subset this entry corresponds to.